src/cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -696,6 +697,73 @@ def load_pem_public_key(self, data):
             None,
         )
 
+    def load_der_private_key(self, data, password):
+        # OpenSSL has a function called d2i_AutoPrivateKey that can simplify
+        # this. Unfortunately it doesn't properly support PKCS8 on OpenSSL
+        # 0.9.8 so we can't use it. Instead we sequentially try to load it 3
+        # different ways. First we'll try to load it as a traditional key
+        key = self._evp_pkey_from_der_traditional_key(data, password)
+        if not key:
+            # Okay so it's not a traditional key. Let's try
+            # PKCS8 unencrypted. OpenSSL 0.9.8 can't load unencrypted
+            # PKCS8 keys using d2i_PKCS8PrivateKey_bio so we do this instead.
+            key = self._evp_pkey_from_der_unencrypted_pkcs8(data, password)
+
+        if key:
+            return self._evp_pkey_to_private_key(key)
+        else:
+            # Finally we try to load it with the method that handles encrypted
+            # PKCS8 properly.
+            return self._load_key(
+                self._lib.d2i_PKCS8PrivateKey_bio,
+                self._evp_pkey_to_private_key,
+                data,
+                password,
+            )
+
+    def _evp_pkey_from_der_traditional_key(self, data, password):
+        mem_bio = self._bytes_to_bio(data)
```

Comment:
If the answer is "that's messier" feel free to punt (but maybe add a TODO)

On Sun, Feb 22, 2015 at 10:14 AM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/backend.py
> <https://github.com/pyca/cryptography/pull/1610#discussion_r25134722>:
>
> > +            key = self._evp_pkey_from_der_unencrypted_pkcs8(data, password)
> > +
> > +        if key:
> > +            return self._evp_pkey_to_private_key(key)
> > +        else:
> > +            # Finally we try to load it with the method that handles encrypted
> > +            # PKCS8 properly.
> > +            return self._load_key(
> > +                self._lib.d2i_PKCS8PrivateKey_bio,
> > +                self._evp_pkey_to_private_key,
> > +                data,
> > +                password,
> > +            )
> > +
> > +    def _evp_pkey_from_der_traditional_key(self, data, password):
> > +        mem_bio = self._bytes_to_bio(data)
>
> Possibly, I think you can rewind a membio.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1610/files#r25134722>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084